src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,712 @@
+	}
+
+	priv = gs_plugin_get_data (plugin);
+
+	if (!gs_flatpak_refine_app (priv->flatpak, app, flags, cancellable, 
+				    error)) {
+		g_debug ("Refining app %s failed!", gs_app_get_id (app));
+		return FALSE;
+	}
+
+	/* We set the state to available because we assume that we can build
+	 * the runtime */
+	if (gs_app_get_state (ext_runtime) != AS_APP_STATE_INSTALLED) {
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
+		gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+	}
+
+	return TRUE;
+}
+
+gboolean
+gs_plugin_add_installed (GsPlugin *plugin,
+			 GsAppList *list,
+			 GCancellable *cancellable,
+			 GError **error)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	return gs_flatpak_add_installed (priv->flatpak, list, cancellable,
+					 error);
+}
+
+gboolean
+gs_plugin_app_install (GsPlugin *plugin,
+		       GsApp *app,
+		       GCancellable *cancellable,
+		       GError **error)
+{
+	g_autofree char *runtime = NULL;
+	g_autofree char *url = NULL;
+	const char *remote_name;
+	GError *internal_error = NULL;
+	GsApp *ext_runtime;
+	GsPluginData *priv;
+
+	/* only process this app if was created by this plugin */
+	if (g_strcmp0 (gs_app_get_management_plugin (app),
+		       gs_plugin_get_name (plugin)) != 0)
+		return TRUE;
+
+	priv = gs_plugin_get_data (plugin);
+
+	ext_runtime = gs_plugin_get_app_external_runtime (plugin, app);
+
+	if (!ext_runtime) {
+		g_debug ("External app '%s' didn't have any asset! "
+			 "Not installing and marking as state unknown!",
+			 gs_app_get_id (app));
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
+
+		return TRUE;
+	}
+
+	gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+
+	if (gs_app_get_state (ext_runtime) == AS_APP_STATE_UNKNOWN) {
+		if (!gs_plugin_build_runtime (plugin, app, ext_runtime, error)) {
+			g_debug ("Failed to build runtime '%s'", runtime);
+			return FALSE;
+		}
+
+		if (!gs_flatpak_refine_app (priv->flatpak, ext_runtime,
+					    GS_PLUGIN_REFINE_FLAGS_DEFAULT,
+					    cancellable, error)) {
+			g_debug ("Failed to refine '%s'",
+				 gs_app_get_id (ext_runtime));
+			return FALSE;
+		}
+	}
+
+	if (gs_app_get_state (ext_runtime) == AS_APP_STATE_UNKNOWN) {
+		g_debug ("Cannot install app asset '%s': state is unknown",
+			 gs_app_get_id (ext_runtime));
+		return FALSE;
+	}
+
+	switch (gs_app_get_state (ext_runtime)) {
+	case AS_APP_STATE_INSTALLED:
+		g_debug ("App asset '%s' is already installed",
+			 gs_app_get_id (ext_runtime));
+		break;
+
+	case AS_APP_STATE_UPDATABLE:
+		g_debug ("Updating '%s'", gs_app_get_id (ext_runtime));
+
+		if (!gs_flatpak_update_app (priv->flatpak, ext_runtime,
+					    cancellable, error)) {
+			g_debug ("Failed to update '%s'",
+				 gs_app_get_id (ext_runtime));
+			return FALSE;
+		}
+		break;
+
+	case AS_APP_STATE_AVAILABLE:
+		g_debug ("Installing '%s'", gs_app_get_id (ext_runtime));
+		gs_flatpak_app_install (priv->flatpak, ext_runtime, cancellable,
+					error);
+		if (gs_app_get_state (ext_runtime) != AS_APP_STATE_INSTALLED) {
```

Comment:
I can believe that, but then you need to make sure that `error` is set in any case when returning `FALSE`. I thoight checking the error from the function above could have been a way to kill two birds with one stone...